### PR TITLE
Supporting automated sip generation for qgsdataitem and qgsvector classes

### PR DESCRIPTION
**Pull request by doublebyte1**
- Removed qgsdataitem.sip and qgsvector.sip from auto_sip.blacklist
 - Annotated qgsdataitem.h for sip
 - regenerated automatically qgsdataitem.sip and qgsvector.sip

## Description
This PR supports the automated generation of the sip files for the QgsVector and QsDataItem classes. In the case of the QgsVector it was not necessary to change anything, apart from removing it from the auto_sip.blacklist. For the case of the QgsVector class, it was necessary to annotate the header with the relevant sip notations (e.g.: SIP_TRANSFERTHIS, SIP_FACTORY, etc).

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by doublebyte1** [comment]
@3nids @m-kuhn Please review the qgsdataitem.h, in case I missed anything :-) Thanks!

File: python/core/qgsvector.sip [hunk 1]
```diff
@@ -1,47 +1,163 @@
...
 
     QgsVector operator+( QgsVector other ) const;
+%Docstring
+ Adds another vector to this vector.
+.. versionadded:: 3.0
+ :rtype: QgsVector
+%End
 
-    QgsVector& operator+=( QgsVector other );
+    QgsVector &operator+=( QgsVector other );
```
Comment (m-kuhn) [inline python/core/qgsvector.sip]: Not strictly required, but can you change the method to

    QgsVector &operator+=( const QgsVector &other );

File: python/core/qgsvector.sip [hunk 2]
```diff
@@ -1,47 +1,163 @@
...
+%End
 
     QgsVector operator-( QgsVector other ) const;
+%Docstring
+ Subtracts another vector to this vector.
+.. versionadded:: 3.0
+ :rtype: QgsVector
+%End
 
-    QgsVector& operator-=( QgsVector other );
```
Comment (m-kuhn) [inline python/core/qgsvector.sip]: Not strictly required, but can you change the method to

QgsVector &operator-=( const QgsVector &other );

File: python/core/qgsvector.sip [hunk 3]
```diff
@@ -1,47 +1,163 @@
...
 
-    QgsVector& operator+=( QgsVector other );
+    QgsVector &operator+=( QgsVector other );
+%Docstring
+ Adds another vector to this vector in place.
+.. versionadded:: 3.0
+ :rtype: QgsVector
+%End
 
     QgsVector operator-( QgsVector other ) const;
```
Comment (m-kuhn) [inline python/core/qgsvector.sip]: Not strictly required, but can you change the method to

QgsVector operator-( const QgsVector &other );

File: python/core/qgsdataitem.sip [hunk 4]
```diff
@@ -1,28 +1,46 @@
...
+ *                                                                      *
+ * Do not edit manually ! Edit header and run scripts/sipify.pl again   *
+ ************************************************************************/
+
+
+
+
+
+
+typedef QgsDataItem *dataItem_t( QString, QgsDataItem * );
```
Comment (m-kuhn) [inline python/core/qgsdataitem.sip]: This isn't needed in python:

    typedef QgsDataItem *dataItem_t( QString, QgsDataItem * ) SIP_SKIP;

File: python/core/qgsdataitem.sip [hunk 5]
```diff
@@ -1,28 +1,46 @@
...
+    if ( sipCpp->inherits( "QgsFavoritesItem" ) )
+      sipType = sipType_QgsFavoritesItem;
+    if ( sipCpp->inherits( "QgsZipItem" ) )
+      sipType = sipType_QgsZipItem;
+    if ( sipCpp->inherits( "QgsDataCollectionItem" ) )
+      sipType = sipType_QgsDataCollectionItem;
+    if ( sipCpp->inherits( "QgsProjectItem" ) )
+      sipType = sipType_QgsProjectItem;
+    else
+      sipType = 0;
```
Comment (m-kuhn) [inline python/core/qgsdataitem.sip]: I think it's a bit faster ot use `qobject_cast` (as before)

**Comment by m-kuhn** [comment]
Thanks, good job!

File: python/core/qgsvector.sip (hunk 1 above)
Comment (doublebyte1) [inline python/core/qgsvector.sip]: done

File: python/core/qgsvector.sip (hunk 2 above)
Comment (doublebyte1) [inline python/core/qgsvector.sip]: done

File: python/core/qgsvector.sip (hunk 3 above)
Comment (doublebyte1) [inline python/core/qgsvector.sip]: done

File: python/core/qgsdataitem.sip (hunk 4 above)
Comment (doublebyte1) [inline python/core/qgsdataitem.sip]: done

File: python/core/qgsdataitem.sip (hunk 5 above)
Comment (doublebyte1) [inline python/core/qgsdataitem.sip]: done

File: python/core/qgsvector.sip (hunk 2 above)
Comment (nyalldawson) [inline python/core/qgsvector.sip]: Actually, this was intentional. QgsVector is small and trivially copiable, so it's faster to pass by value then reference here (add elsewhere in the class)

Comment (m-kuhn) [inline python/core/qgsvector.sip]: can we merge it with the `const QgsVector other` state (i.e. without the reference `&` part in there)?

Comment (nyalldawson) [inline python/core/qgsvector.sip]: Sure

**Comment by 3nids** [comment]
I updated the sipify script. You might need to rebase again, sorry!

**Comment by doublebyte1** [comment]
@3nids Ok. I updated the source code and pushed the changes. The build compiles now! :-)

**Comment by 3nids** [comment]
manually merged, thanks a lot

**Comment by 3nids** [comment]
I am really sorry the commits were transferred to me as author. I had troubles with the merging and did not do it properly apparently. really sorry for that.

**Comment by 3nids** [comment]
nope, it's ok, it's just the merge commits
https://github.com/qgis/QGIS/commits?author=doublebyte1

**Comment by doublebyte1** [comment]
@3nids great, I am glad I could help!

**Pull request merged**